### ISSUE TITLE
"could not find key" errors and an assertion failure with pg_regress tests

### ISSUE THREAD

**Issue by hlinnaka** [neondatabase/neon#2693]
Saw some interesting failures with https://github.com/neondatabase/neon/pull/2632. I beileve they are not related to the changes in the PR, but I'm not 100% sure.

https://github.com/neondatabase/neon/actions/runs/3321045337/jobs/5488310468

From pg.log:
```
2022-10-25 13:06:55.795 GMT [15990] ERROR:  could not read relation size of rel 1663/16384/16996.0 from page server at lsn 0/01696840
2022-10-25 13:06:55.795 GMT [15990] DETAIL:  page server returned error: could not find data for key 000000067F000040000000426400FFFFFFFF at LSN 0/1696070, for request at LSN 0/1ADB668
2022-10-25 13:06:55.795 GMT [15990] STATEMENT:  create index test_range_spgist_idx on test_range_spgist using spgist (ir);
...
2022-10-25 13:06:56.131 GMT [15990] ERROR:  could not read relation size of rel 1663/16384/17008.0 from page server at lsn 0/01696840
2022-10-25 13:06:56.131 GMT [15990] DETAIL:  page server returned error: could not find data for key 000000067F000040000000427000FFFFFFFF at LSN 0/1696070, for request at LSN 0/1BB2540
2022-10-25 13:06:56.131 GMT [15990] STATEMENT:  create table test_range_excl(
	  room int4range,
	  speaker int4range,
	  during tsrange,
	  exclude using gist (room with =, during with &&),
	  exclude using gist (speaker with =, during with &&)
	);
...
2022-10-25 13:07:08.825 GMT [16444] ERROR:  could not read relation size of rel 1663/16384/20183.0 from page server at lsn 0/01696840
2022-10-25 13:07:08.825 GMT [16444] DETAIL:  page server returned error: could not find data for key 000000067F0000400000004ED700FFFFFFFF at LSN 0/1696070, for request at LSN 0/4CE8B40
2022-10-25 13:07:08.825 GMT [16444] STATEMENT:  CREATE INDEX concur_appclass_ind on concur_appclass_tab
	  USING gist (i tsvector_ops (siglen='1000'), j tsvector_ops (siglen='500'));
2022-10-25 13:07:08.853 GMT [16444] ERROR:  could not read relation size of rel 1663/16384/20184.0 from page server at lsn 0/01696840
2022-10-25 13:07:08.853 GMT [16444] DETAIL:  page server returned error: could not find data for key 000000067F0000400000004ED800FFFFFFFF at LSN 0/1696070, for request at LSN 0/4CE8B40
2022-10-25 13:07:08.853 GMT [16444] STATEMENT:  CREATE INDEX concur_appclass_ind_2 on concur_appclass_tab
	  USING gist (k tsvector_ops (siglen='300'), j tsvector_ops);
```

and finally:
```
TRAP: FailedAssertion("((PageHeader) (page))->pd_special >= SizeOfPageHeaderData", File: "/__w/neon/neon//vendor/postgres-v14/src/include/storage/bufpage.h", Line: 317, PID: 17467)
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(ExceptionalCondition+0xbd)[0x55df70e8d592]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0xf1777)[0x55df7080c777]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(ginHeapTupleFastInsert+0x3d1)[0x55df7080d287]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(gininsert+0x1c3)[0x55df7081549c]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(index_insert+0x17b)[0x55df7087f170]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(ExecInsertIndexTuples+0x30b)[0x55df70a9cd67]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x3ca6e0)[0x55df70ae56e0]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x3ccd05)[0x55df70ae7d05]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x38ef6d)[0x55df70aa9f6d]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x38356d)[0x55df70a9e56d]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x385f5b)[0x55df70aa0f5b]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(standard_ExecutorRun+0x1ea)[0x55df70a9ec14]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(ExecutorRun+0x59)[0x55df70a9ea27]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x5d9ea4)[0x55df70cf4ea4]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x5db83e)[0x55df70cf683e]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(PortalRun+0x307)[0x55df70cf5dea]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x5d463d)[0x55df70cef63d]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(PostgresMain+0x82e)[0x55df70cf3e61]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x50ee77)[0x55df70c29e77]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x50e7a9)[0x55df70c297a9]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(+0x50a98b)[0x55df70c2598b]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(PostmasterMain+0x1286)[0x55df70c251d8]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(main+0x37c)[0x55df70b22c68]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xea)[0x7f73e7994d0a]
postgres: cloud_admin regression 127.0.0.1(43574) INSERT(_start+0x2a)[0x55df707dc76a]
2022-10-25 13:07:26.384 GMT [15733] LOG:  server process (PID 17467) was terminated by signal 6: Aborted
2022-10-25 13:07:26.384 GMT [15733] DETAIL:  Failed process was running: insert into gin_test_tbl select array[1, 2, g] from generate_series(1, 20000) g;
```

**Comment by knizhnik** [comment]
> @knizhnik I suspect this is a bug with [neondatabase/postgres#224](https://github.com/neondatabase/postgres/pull/224). That was merged to the `main` (i.e. v14) branch in the `postgres` repository, but we have _not_ yet updated the submodule reference in the `neon` repository. I can reproduce this failure when I run the pg_regress tests on my laptop, with fresh checkout of `main` under vendor/postgres-v14.

Sorry, my mistake: overflow in upper bond calculation caused by REL_METADATA_PSEUDO_BLOCKNO.
This is why I didn't wan to change this code.
I have created two more PRs.

**Comment by hlinnaka** [comment]
This was fixed in commit neondatabase/postgres@e219c501f01739f239247dafb68e30ed1e2f9b4e (v15) and neondatabase/postgres@fb475c7d5d1bba44f8693a0f3a1a0e784b740ba6 (v14)
